src/util/code_observer.h

```diff
@@ -0,0 +1,161 @@
+/*******************************************************************\
+
+Module: Misc Utilities
+
+Author: Kurt Degiorgio
+
+\*******************************************************************/
+
+/// \file
+/// code_observer.h
+
+/// This file contains a couple of utility classes that should be used when one
+/// needs to modify or read all codet's of a particular type from a given symbol
+/// table.  This saves you from having to manually iterate through the symbol
+/// table.  These classes essentially implement the observable pattern.
+///
+/// In general, you first register one or more observers with either
+/// code_observert or const_code_observert. The observer is a callback that is
+/// associated with a specific irep_idt. This is used to determine for which
+/// codet's the observer will be notified. The second step involves calling
+/// visit_symbols which will in turn iterate through every codet inside the
+/// given symbol-table and notify registered observers only if the id of the
+/// codet in question matches the associated id of the observer.
+///
+/// Example: Replacing all codet(ID_function_call) with codet(ID_skip)
+///
+/// \code
+///
+/// void A::replace_function_calls(codet &code)
+/// {
+///   PRECONDITION(code.get_statement()==ID_function_call);
+///   code = codet(ID_skip);
+/// }
+///
+/// ...
+///
+/// code_observert code_observer;
+/// code_observer.registerObserver(ID_function_call,
+///     &A::replace_function_calls,
+///     this,
+///     std::placeholders::_1);
+/// code_observer.visit_symbols(symbol_table);
+///
+/// ...
+///
+/// \endcode
+///
+/// Note: if the callback only needs to read (and not modify) the codet, then
+/// const_code_observert should be used otherwise use code_observert.
+///
+/// Note': each unique irep_idt can only be associated with one observer.
+
+#ifndef CPROVER_UTIL_CODE_OBSERVER_H
+#define CPROVER_UTIL_CODE_OBSERVER_H
+
+#include <map>
+#include <functional>
+#include <type_traits>
+
+#include "symbol_table.h"
+#include "std_code.h"
+
+/// \brief base class for all code observer implementations
+template<class T> class code_observer_baset
+{
+static_assert(std::is_base_of<codet, T>::value, "T is not derived from codet");
+public:
+  virtual ~code_observer_baset() {}
+
+  /// registers an observer, that is associated with specific codet id.
+  /// \param id: id of the codet to be registered
+  /// \param f: observer callback, that is associated with the given, \p id
+  /// \param args: arguments associated with the given callback, \p f
+  template<class F, class ...Args>
+    void registerObserver(const irep_idt &id, const F &&f, Args&&... args)
```

Comment:
I don't think we indent here. No CaMelcaSe, please.

Reply:
done